Update Accordion design

chris-gds opened the pull request:
## What

Update the Design of the Accordion component to improve accessibility.
 
## Why

The existing iteration of the Accordion: the section headings in accordions can be mistaken for links, but are treated as buttons - this fails [WCAG 2.1 success criterion 1.3.1 Info and Relationships
](https://www.w3.org/TR/UNDERSTANDING-WCAG20/content-structure-separation-programmatic.html) this update addresses this.

Also, there was initially 3 different Accordion Designs on different parts of GOV.UK Publishing. This impacted [WCAG 3.2.4](https://www.w3.org/WAI/WCAG21/Understanding/consistent-identification.html) as the differences were not consistent they were ultimately confusing for users.

(This iterative update within the Design System (GOV.UK Frontend) will also be updated on [GOV.UK Publishing Gem](https://components.publishing.service.gov.uk/component-guide/accordion))

## Visuals 
(Before: Left, After: Right)

https://github.com/alphagov/govuk-design-system/issues/1706

## Browser testing
- [x]  IE8
- [x]  IE9
- [x]  IE10
- [x]  IE11
- [x] Edge
- [x] Chrome
- [x] Chrome 400% zoom
- [x] Firefox
- [x] Firefox 300% zoom text only 
- [x] Safari
- [x] Safari (iOS)
- [x] Chrome (iOS)
- [x] Chrome (Android)
- [x] Samsung Internet (Android)

<details>
  <summary>Selected Results:</summary>

(Trigger warning: GIFs to follow [do not pause after 3 loops](https://www.w3.org/TR/WCAG20-TECHS/G152.html))

IE8
![IE8](https://user-images.githubusercontent.com/71266765/128702424-0c8b1e72-109a-40da-b613-fd31d769fcc8.gif)

IE9
![IE9](https://user-images.githubusercontent.com/71266765/128702430-5f8353bd-27a0-4e46-ba66-314856b290f5.gif)

IE10
![IE10](https://user-images.githubusercontent.com/71266765/128702434-d37ef672-fdc9-45a1-84e4-513ed66cb21f.gif)

IE11
![IE11](https://user-images.githubusercontent.com/71266765/128702438-8635cbe8-bb71-4e79-9d6c-e900eeb4713a.gif)

Edge

![Edge](https://user-images.githubusercontent.com/71266765/128703032-58402769-c9cc-4923-87f6-f4d5837a3b3a.gif)

Chrome zoom

https://user-images.githubusercontent.com/71266765/128702962-93db59b9-5a33-4675-9a95-5bacc5748d58.mov

Firefox Zoom-text only

https://user-images.githubusercontent.com/71266765/128702994-06028a6b-2f78-499b-a170-6f4e97f305dc.mov

Android Pixel 5 Chrome

![pixel-chrome](https://user-images.githubusercontent.com/71266765/128709525-38f52349-08a1-4c9f-9747-110bd5b700a0.gif)

Android Samsung 20 Chrome

![samsung-20-chrome](https://user-images.githubusercontent.com/71266765/128709613-37b2246f-249a-4470-aa5b-16d1f310ba8b.gif)

Android Samsung 20 Samsung Internet

![samsung-20-samsung](https://user-images.githubusercontent.com/71266765/128709652-409cd3bc-8086-4617-9fde-4f27199312b0.gif)

Safari macOS Catalina

![safari-13 1-macOS-catalina](https://user-images.githubusercontent.com/71266765/128703210-c8dac43a-0903-4ad5-9d2c-b1c5eb45da2e.gif)

Safari macOS BigSur

![safari-macOS-bigSur](https://user-images.githubusercontent.com/71266765/128703216-48ec07c2-afa8-4653-baca-2c5f6ad6e74c.gif)

iPadPro
![iOS-iPadPro11-2020](https://user-images.githubusercontent.com/71266765/128702539-899ac391-374b-4e35-8af9-00d29b23afd2.gif)

iPhone 11 Pro Max
![iOS-iPhone11ProMax](https://user-images.githubusercontent.com/71266765/128702544-c8fe9495-088c-4da6-9dce-d2b010ba8d88.gif)

iPhone 12
![iOS-iPhone12](https://user-images.githubusercontent.com/71266765/128702551-eb67fcbe-de49-4eb3-8d16-6a184a59d830.gif)

iPhone SE 2020
![iOS-iPhoneSE2020](https://user-images.githubusercontent.com/71266765/128702558-3903fb68-573e-487c-ae27-dd696f478649.gif)

Chrome - iPhone 12
![chrome-ios](https://user-images.githubusercontent.com/71266765/128718159-3376ea1f-3564-494a-a603-6318bc2a11d4.gif)


</details>

## Assistive Technology testing
- [x] JavaScript Disabled
- [x] CSS Disabled
- [x] CSS + JS Disabled
- [x] Voiceover / Safari (macOS Big Sur)
- [x] Voiceover / Safari (iOS 14.6)
- [x] NVDA 2020.3 / Firefox 88
- [x] NVDA 2020.3 / Chrome
- [x] JAWS 2020 / Chrome
- [x] JAWS 2020 / IE11
- [x] Windows Magnifier
- [x] Firefox (User preferred colours)
- [x] Chrome - Blurred vision, protanopia, deuteranopia, tritanopia and achromatopsia
- [x] Chrome (User preferred Colours (Increase contrast, Greyscale, Inverted colour, Inverted Greyscale, Yellow on black))
- [x] Windows High Contrast Modes - Chrome, IE, Edge

## Known issues

Untested
- [x] Talkback / Chrome 86 (Android 10)  
- [x] Dragon (speech recognition) \| 15 or later \| Internet Explorer 11*
- [x] Voiceover / Safari (iOS 12.1)

<details>
  <summary>Results:</summary>
  
JavaScript Disabled

![no-JS](https://user-images.githubusercontent.com/71266765/128702817-ff0e4e59-c05c-4fef-b24d-a938286f261d.png)

CSS Disabled

![no-CSS](https://user-images.githubusercontent.com/71266765/128702775-c94b1c22-d6a1-4d59-809a-aeeec60f098c.png)

CSS + JS  Disabled

![no-CSS-no-JS](https://user-images.githubusercontent.com/71266765/128702798-53bfd42b-b0fa-4aaa-b16b-7f08a49bc62f.png)

VoiceOver Safari

https://user-images.githubusercontent.com/71266765/128705909-0bc64625-b261-4c31-9c99-c021017fd49e.mov

VoiceOver iOS (14.6)*
*`main` landmark present in testing environment

https://user-images.githubusercontent.com/71266765/128704394-1677774e-ce8f-465d-8507-09a5a8be04d7.mp4

NVDA 2020.4 / Firefox 88

https://user-images.githubusercontent.com/71266765/128702866-fca26121-1966-4c1a-8a97-988c366d5696.mov

NVDA 2020.4 / Chrome

https://user-images.githubusercontent.com/71266765/128702849-3773d32e-2bc1-4f6e-b254-c70352cd5b4a.mov

JAWs / Chrome  

https://user-images.githubusercontent.com/71266765/128719445-c18b226a-8819-4f37-8e79-39761c8fec09.mov

JAWs / IE  

https://user-images.githubusercontent.com/71266765/128702743-7bba985d-1290-49de-8030-e41079e0717c.mov

Chrome (Blurred vision, protanopia, deuteranopia, tritanopia and achromatopsia)

https://user-images.githubusercontent.com/71266765/128707921-d4182bad-0597-4509-a2aa-5770f520a654.mov

Chrome (Increase contrast, Greyscale, Inverted colour, Inverted Greyscale, Yellow on black)

https://user-images.githubusercontent.com/71266765/128708277-2eb79045-8b67-4c05-9b8c-9273338138df.mov

Windows Magnifier

![Windows-magnifier](https://user-images.githubusercontent.com/71266765/128721094-6cf6e86f-423a-4aa3-bf02-97703d1928d3.gif)

Chrome (Windows High Contrast Modes)

https://user-images.githubusercontent.com/71266765/128722205-50df7a61-5660-4d8e-a024-a1e2458d2a7d.mov

Edge (Windows High Contrast Mode)

https://user-images.githubusercontent.com/71266765/128721473-748f30a9-20ce-48c8-87bd-c16cce52e439.mov

IE11 (Windows High Contrast Mode)

![Windows-high-contrast](https://user-images.githubusercontent.com/71266765/128703219-10857df0-ee31-449e-8915-166561eb7aed.gif)


</details>

## Anything else

**Breaking Change** as the [ID has now been removed from the summary element within the markup](https://github.com/alphagov/govuk-frontend/pull/2257/commits/fa572d331f030685df53923951cc5e0286b9c863)

### User preferred colours

[User's may adjust their colour schemes](https://accessibility.blog.gov.uk/2017/03/27/how-users-change-colours-on-websites/).  Testing with different colour schemes revealed some problems, having nested elements within a button [required additional adjustments](https://github.com/alphagov/govuk-frontend/commit/f258ce4e3cddc653e502193bedd32dab9a44787a) in order to improve the experience. 

There is further conversation about this use and [an opinion piece can be found here](https://kilianvalkhof.com/2021/css-html/prefers-contrast-forced-is-a-mistake/)
https://github.com/w3c/csswg-drafts/issues/5433#issuecomment-785251576

### NVDA fix

<details>
  <summary>Testing videos of NVDA</summary>

Before - the sections are broken up as user navigates.
Within reader-mode the user requires multiple keypresses to navigate, also the audio is heard in an disjointed manner.

![128227998-ea003a53-9492-4a86-863c-5303950a8783](https://user-images.githubusercontent.com/71266765/128982685-9d650c0f-ba5b-432a-be22-e8b389d942dd.gif)

After - read out as one whole block.

https://user-images.githubusercontent.com/71266765/128356759-b3be940f-bfd2-4aff-85cf-807a4e3a8807.mov

</details>

<details>
  <summary>Additional Details and screenshots:</summary>

The issue before:
![before](https://user-images.githubusercontent.com/71266765/123416396-662a3500-d5ae-11eb-94bf-fba1dbcbbf37.gif)

After:

![firefox-color2](https://user-images.githubusercontent.com/71266765/128706575-87caa406-86cf-43a7-8fad-3678d045750a.gif)

![firefox-color1](https://user-images.githubusercontent.com/71266765/128706587-2c704e37-2322-479c-8c11-6f8b0b918de7.gif)

---

To replicate:

`Firefox > preferences > Language and appearance > Colours`:
  
Update to high contrast scheme
![Screenshot 2021-06-22 at 11 02 43](https://user-images.githubusercontent.com/71266765/123416612-b1dcde80-d5ae-11eb-90e3-4da2f86bdd83.png)

Update to reduced contrast scheme

![image](https://user-images.githubusercontent.com/71266765/123416668-c6b97200-d5ae-11eb-8e59-6912788ba13d.png)

</details>

## Details of breaking change

- which users are affected: TBC
- the change that’s been made: TBC
- changes users will have to make: TBC
- impact of users not making those changes: TBC

chris-gds commented on src/govuk/components/accordion/accordion.js:
```diff
@@ -93,49 +105,55 @@ Accordion.prototype.initSectionHeaders = function () {
+    // Add all attributes but not ID as this is being add directly on $wrapperFocusHeading
+    if (attr.nodeName !== 'id') {
+      $button.setAttribute(attr.nodeName, attr.nodeValue)
+      console.log(attr.nodeName, attr.nodeValue)
```
🙈

36degrees commented on src/govuk/components/accordion/accordion.js:
```diff
@@ -93,49 +105,55 @@ Accordion.prototype.initSectionHeaders = function () {
 
   // Copy existing span element to an actual button element, for improved accessibility.
   var $button = document.createElement('button')
-  $button.setAttribute('type', 'button')
```
This is setting the button's`type`, not `role`.

I think there's _potential_ for this change to cause issues for accordions inside forms, as the default `type` for a button is `submit` and I can't see anything in the JS to prevent the default click behaviour on the button which would be to submit the form.

frankieroberto replied:
Yes, agreed – this would be a breaking bug for anywhere accordions are used inside forms. My service current does this, although we’re in the process of moving away from using the accordion in this way.

chris-gds replied:
All makes sense, restored. 🙌🏼

hannalaakso commented on src/govuk/components/accordion/accordion.test.js:
```diff
@@ -118,13 +118,24 @@ describe('/components/accordion', () => {
+
+          const numberOfExampleSections = 2
+
+          const numberOfIcons = await page.evaluate(() => document.body.querySelectorAll('.govuk-accordion .govuk-accordion__section .govuk-visually-hidden').length)
```
It would be good to check that the visually hidden div contains the " this section" text. Similar to what we do with the "Show all sections" button.

hannalaakso commented on src/govuk/components/accordion/accordion.test.js:
```diff
@@ -118,13 +118,24 @@ describe('/components/accordion', () => {
+        })
+      })
+
+      describe('hidden text', () => {
```
It would be great if we could add a test for the "Show" text in the button as well.

chris-gds replied:
Theres's now a test for the text in it's entirely - hopefully enough?

frankieroberto commented on src/govuk/components/accordion/_index.scss:
```diff
@@ -42,76 +35,214 @@
+
+      width: 100%;
+
+      padding: govuk-spacing(2) 0 govuk-spacing(6);
```
Could the padding at the bottom of the button be reduced, to say `govuk-spacing(2)` to match the padding on top (10px instead of 30px)? The padding on the top of the expanded content section could then be increased slightly to maintain the spacing.

The rationale for reducing the padding would be to reduce the height that each section takes up when collapsed, as this new design is significantly taller than the current one. This might impact usability, as it is harder to scan a list of sections when you can't see as many of them at once.

The version on the [Coronavirus page](https://www.gov.uk/coronavirus) currently has 96px high sections whereas this version has 107px high sections. Existing component has 61px high sections.

chris-gds replied:
👋🏼  Hey @frankieroberto, so at this stage this current Design has been reviewed and finalised by multiple teams however the observations you've raised would be worth adding to user-testing to see the impact.

> as this new design is significantly taller than the current one. This might impact usability, as it is harder to scan a list of sections when you can't see as many of them at once.

Also, there is a [condensed version](https://components.publishing.service.gov.uk/component-guide/accordion/condensed_layout/preview) of the layout that exists for GOV.UK that might be of some interest.  I'm aware that the Accordion's guidance centres around this element being used a "last resort" and _if_ the Accordion houses too much Content then it _might_ be worth a review

frankieroberto replied:
Fair enough! Will the condensed version be an option within the component?

chris-gds replied:
AFAIK not within the Design System however this option will persist on GOV.UK.  However, there _might_ be more of a case to remove this as well.

I suppose it prompts an upstream question, if there are lot of sections and the larger Design potentially impacts how a user scans  - why are there that many sections?  Does the Content need review? Also discovering the details of the scanning.. impact.
 
I've be curious to discover all the instances of this being used cross-gov to get a better understanding of use.  Equally the reserve could be true + once this is in situ / user research this might be adjusted.

frankieroberto replied:
@chris-gds I suspect the accordion is used (abused?) in all sorts of different ways!

One of the primary ways I’ve seen it used though is for navigation, where each section is essentially a list of links instead of content in paragraphs.

Examples:

* https://explore-education-statistics.service.gov.uk/find-statistics
* https://www.ethnicity-facts-figures.service.gov.uk/education-skills-and-training

Perhaps there’s better ways to do this though.

chris-gds replied:
>I suspect the accordion is used (abused?) in all sorts of different ways!

+1, thanks for these - it's useful to see how it's being used elsewhere (I've not seen these before)
Out of scope to comment on this use within this PR but having more examples is useful to understand how this update might interact

hannalaakso replied:
Thanks @frankieroberto @chris-gds for your thoughts!

I think you make a really good point about accordions often being misused when another component/pattern might be more suitable. We have actually have it in our backlog to update the component guidance to be clearer on this point when we publish this new version. Have a look [here](https://github.com/alphagov/govuk-design-system/issues/1684#issuecomment-846946557) (it's point 2) and feel free to leave any comments on that issue if you think there is anything else we should take into account when updating the guidance.

frankieroberto replied:
@hannalaakso thanks, added a comment to that issue. Whether accordions are being misused or not is a thorny debate!

chris-gds replied:
In response - @christopherthomasdesign feedback:

> The component being taller generally is a necessary side effect of having the show all controls on a new line. I actually agree with frankie on making the padding equal on the top and bottom – need to agree this with designers on gov.uk because there was deffo a reason for it Nikin happy with 15px rather than 30 under each row

Design @nnagewad being altered to include this change.

chris-gds commented on src/govuk/components/accordion/accordion.js:
```diff
@@ -44,30 +51,36 @@ Accordion.prototype.init = function () {
-  this.$openAllButton.setAttribute('type', 'button')
+
+  // Create icon, add to element
+  var icon = document.createElement('span')
```
Makes sense, this should be reflected now

lfdebrux commented:
Browser testing - Edge - Windows High Contrast Mode

https://user-images.githubusercontent.com/503614/127344634-8907dc4f-dc12-430f-a2c9-bdd88da1c5f5.mp4

lfdebrux commented on src/govuk/components/accordion/accordion.test.js:
```diff
@@ -118,13 +116,46 @@ describe('/components/accordion', () => {
+
+        expect(ShowOrHideButtonText).toEqual('Hide this section')
+      })
+
```
There's a bit of a (pre-existing) gap in the tests, currently there are no tests for an accordion with additional description/summary text.

Given that this PR adds a bit more sophisticated handling of the additional description, it might be worth adding tests such as `it('should move the additional description to the button text')`; would that be okay?

chris-gds replied:
Make sense, i've added a tests to check for the span being changed. However it might be worth filling that gap to add the missing check of the summary text just existing.

hannalaakso commented on src/govuk/components/accordion/accordion.test.js:
```diff
@@ -118,13 +116,46 @@ describe('/components/accordion', () => {
+        })
+      })
+
+      it('should change the (combined) Show[ this section] text to Hide[ this section ] when sections are opened', async () => {
```
This test should not pass, but is passing because the accordion is still open from another test scenario.

We should possibly be scoping the tests differently in general, @lfdebrux will investigate this and will raise a PR against main branch to fix.

chris-gds replied:
As a backup to that action - updated the test and added a reset as well to tackle this

hannalaakso commented on src/govuk/components/accordion/accordion.js:
```diff
@@ -50,7 +50,6 @@ Accordion.prototype.init = function () {
   }
```
As moving the summary inside the button is a slightly unexpected change (which then requires the change in this commit), we discussed that it would be good to 
- move code that moves the summary inside the button into this commit to have a self contained commit of all the related changes
- add a bit more detail to the commit message about why we’re moving the summary inside the button (mainly to do with design as we didn’t want the summary text below the button element)
- why moving the summary means that it can no longer be a span

hannalaakso replied:
A small thing but it seems that this comment from https://github.com/alphagov/govuk-frontend/pull/2257/commits/408b8cdd85e7abcd2a2152f14e130d1cad263815

> (@nnagewad) "We’ve decided to move the summary in order to better visually illustrate the additional context of accordion heading by placing it directly beneath the heading as opposed to placing it directly beneath the reveal/expand button"[4]

should probably be in https://github.com/alphagov/govuk-frontend/pull/2257/commits/61020eb0de84361ff2a2b9da3dc0ca53d106b223. 

We could also signpost to https://html.spec.whatwg.org/multipage/form-elements.html#the-button-element re: 
> Content model: 
> [Phrasing content](https://html.spec.whatwg.org/multipage/dom.html#phrasing-content-2)

chris-gds replied:
Moved the justification into the other commit & added the additional sign post to phrasing.

hannalaakso commented on src/govuk/components/accordion/_index.scss:
```diff
@@ -190,77 +252,57 @@
-    .govuk-accordion__icon:after {
-      height: 100%;
+    // Remove the bottom margin from the last item inside the content
+    .govuk-accordion__section-content > :last-child {
```
We moved this rule here but didn't delete the same rule from line 32, the duplicate should be removed.

hannalaakso commented on src/govuk/components/accordion/_index.scss:
```diff
@@ -263,6 +263,15 @@
       &:active {
         background: $govuk-link-active-colour;
       }
+
```
We discussed that this commit https://github.com/alphagov/govuk-frontend/pull/2257/commits/4d44935136f9470d851fb6f7e6e37998f3980176# 
might need 
- a bit more detail on the original problem that we're fixing
- the desired outcome
- to be split in two as there are two slightly different changes here

It would also be good to dig out the original bug report for NVDA - @chris-gds let me know if I can help with that.

chris-gds replied:
Agreed, updating commit with details splitting out commit thank contains pause 

-----

Additional Details for context 
Original issue flagged:

Not sure if this is a quirk of NVDA, but if you navigate through the accordion (arrow keys), it reads out the heading (together with ‘Show’) and ‘this section’ as 2 separate items, which could be confusing:

_clickable  heading    level 2  button  collapsed    Know your audience,Show
heading    level 2  button  collapsed    this section_

---- 

Clinic Notes:
"Notably, we didn’t take issue with when NVDA split the button into separate blocks. It’ll be good regardless to get some detailed rationale as to why this could be acceptable and how theoretically impactful the lack of pausing is.

The rationale was NVDA has been seen to treat headings / buttons in a similar fashion elsewhere on the web. There would be a degree of expectation that the user might be used to this."

---

Action -  agreed to fix issue.

chris-gds replied:
- Added more detail to commits about the history of being spotted, feedback, investigation and fix also,
- Added inline comments to further outline issue.
- PR also updated with before and after videos for visual reference of the issue
- Pause split out of original commit for clarity

hannalaakso commented on src/govuk/components/accordion/accordion.test.js:
```diff
@@ -118,13 +116,46 @@ describe('/components/accordion', () => {
+        })
+      })
+
+      describe('hidden comma', () => {
```
It would be good to update this test so that we can check that the div content equals the comma symbol.

hannalaakso commented on src/govuk/components/accordion/_index.scss:
```diff
@@ -44,74 +37,212 @@
+      vertical-align: text-top;
+
+      // IE8 fallback of icon
+      @include govuk-if-ie8 {
```
The arrow lines for IE8 are very narrow (might not be visible to all users) and also slightly overlaps the borders of the box. We don't need to do anything too fancy here for IE8 but there might some slightly improvements we could make, I think @chris-gds is going to experiment with bolding the icon.

chris-gds replied:
Update to alter icon to improve this (Before / After) 

![IE8-update](https://user-images.githubusercontent.com/71266765/128318304-6df191b3-699f-4c53-8213-5983b76f437b.jpg)

(Possible action point: Design icon AC: include IE8)

hannalaakso commented on src/govuk/components/accordion/accordion.js:
```diff
@@ -104,6 +119,18 @@ Accordion.prototype.initHeaderAttributes = function ($headerWrapper, index) {
+  var $icon = document.createElement('span')
+  $icon.classList.add(this.upChevronIconClass)
+  $showIcons.appendChild($icon)
+  $wrapperShowHideIcon.classList.add(this.sectionShowHideTextClass)
```
Only a small thing but the variable name is slightly confusing as it contains the toggle text.

chris-gds replied:
Updated the naming to be clearer to `wrapperToggleText`

hannalaakso commented on src/govuk/components/accordion/accordion.test.js:
```diff
@@ -118,13 +116,46 @@ describe('/components/accordion', () => {
           await page.goto(baseUrl + '/components/accordion/preview', { waitUntil: 'load' })
 
           const numberOfExampleSections = 2
+          const numberOfIcons = await page.evaluate(() => document.body.querySelectorAll('.govuk-accordion .govuk-accordion__section .govuk-accordion__section-toggle-text .govuk-visually-hidden').length)
```
Not a big deal but a slightly confusing variable name here.

hannalaakso commented on src/govuk/components/accordion/_index.scss:
```diff
@@ -190,77 +252,57 @@
-    @media (hover: none) {
-      .govuk-accordion__section-button:hover {
-        text-decoration: none;
+      &:active {
```
This is probably redundant as the container is a div.

hannalaakso commented on src/govuk/components/accordion/_index.scss:
```diff
@@ -87,6 +89,73 @@
+
+        transform: rotate(-45deg);
+
+        border-top: 2px solid;
```
A small thing but we could explicitly set the colour for the borders, at the moment it's being inherited from the toggle element which could change.

hannalaakso reviewed (commented):
Thanks @chris-gds for all your work on this, it's looking good 🎉 

I left some comments, mainly to do with adding some more code comments and clarifying a few things.

A couple of other things that we spotted:
- [x] iOS Voiceover announces summary text twice on 12.1
- [ ] iOS Voiceover announces the show all button as a main landmark - I think you correctly said @chris-gds that this is to do with the context we're testing in. It would be good to see up a full page example with the accordion which should fix this, let me know if I can help with that @chris-gds 
- [x] NVDA read mode (Insert + Arrow Up/Down) splits the heading/summary/button/whole element into separate entities, not sure if this is an issue for users but I think you said @chris-gds that you might know why this is and will investigate
![nvda](https://user-images.githubusercontent.com/5007934/128227998-ea003a53-9492-4a86-863c-5303950a8783.gif)
- [x] Picked up as part of the design review: the headings should be the same as `govuk-heading-m` and the summary should be the same as `govuk-body`
- [x] We know that there are some really useful contextual comments for the design changes [here](https://docs.google.com/document/d/1CG0OUT1tzy7q9juK4O0Iv974854hiBK5GMGF9_0ma08/edit#heading=h.1i6pnax09yoi) - hopefully we could move those into the commit messages to explain the design rationale 
- [x] There are a couple of things we could move around in the commit messages to make it easier to review the changes - listed [here](https://docs.google.com/document/d/1CG0OUT1tzy7q9juK4O0Iv974854hiBK5GMGF9_0ma08/edit#heading=h.i43eyiccsig0), but our team could also pick those up later

chris-gds commented on src/govuk/components/accordion/_index.scss:
```diff
@@ -44,74 +37,212 @@
+
+      position: relative;
+
+      width: 20px;
```
Agreed - the `govuk_publishing_components` version of this iteration contains relative values[1] in order to scale as expected, in the context of this contribution it was noted this pattern is perhaps not present within the Design System at present so didn't wish to introduce a change in convention.  This could be worth exploring at a later date for all icons to meet a zoom text-only AC.

[1]https://github.com/alphagov/govuk_publishing_components/commit/192ff22f1332c18d710e42889431b6c92eeda3e1

chris-gds replied:
After some thought decided it was more favourable to introduce relative values in this context. [This has been applied here:](https://github.com/alphagov/govuk-frontend/pull/2257/commits/aae5ab9486383189d1450057d1f49bd0e18579e8)

Updated visuals (zoom text only firefox):

![Aug-09-2021 10-58-59](https://user-images.githubusercontent.com/71266765/128689804-37c00076-3309-4bd4-8221-62d24010d36a.gif)

hannalaakso replied:
Brilliant, thanks @chris-gds 🎉

hannalaakso commented on src/govuk/components/accordion/accordion.js:
```diff
@@ -93,49 +105,78 @@ Accordion.prototype.initSectionHeaders = function () {
 
+  // If summary content exists add to DOM in correct order
   if (typeof ($summary) !== 'undefined' && $summary !== null) {
-    $button.setAttribute('aria-describedby', this.moduleId + '-summary-' + (index + 1))
```
If we remove this, we're no longer linking the button to the summary text. That might be okay as the summary is now inside the button but we should probably re-test on screen readers to see if the summary is still announced -  I think you were anticipating this @chris-gds?

If we do retain this change, we should also delete the id from the summary, I don't think it serves any other purpose.

chris-gds replied:
>we're no longer linking the button to the summary 

Exactly, as the summary text is now _inside_ the button the additional `aria-describedby` is not longer required as it's not needed to be referenced by something outside.  The heading + summary naturally is announced should be present.

> re-test on screen readers to see if the summary is still announced - I think you were anticipating this @chris-gds?

Yes, there's a fresh round of testing that's needed to just make sure this works as intended.

> If we do retain this change, we should also delete the id from the summary, I don't think it serves any other purpose.

This is a good point 🤔  - as we're updating the template (mentioned below might need a talk) to add `govuk-heading-` this can be removed as well.

chris-gds commented:
> A couple of other things that we spotted:
> * [x]  iOS Voiceover announces summary text twice on 12.1

Removed `aria-describedby` to resolve this -  as the summary has moved inside the button this additional aria attribute is no longer needed 🎉  so now prevents the duplicated announcement.

> * [ ]  iOS Voiceover announces the show all button as a main landmark 

I'm 90% certain this due to the testing environment being wrapped in a `main` as this also announced on other elements like breadcrumbs. Will confirm during deeper testing once accordion is placed on a different page.

> * [x]  NVDA read mode (Insert + Arrow Up/Down) splits the heading/summary/button/whole element into separate entities, 

The fix wasn't in place correctly, this has been resolved.

> * [x]  Picked up as part of the design review: the headings should be the same as `govuk-heading-m` and the summary should be the same as `govuk-body`

This has been updated however it's worth noting as a result this is a **breaking change** as it update the template file. 

Also if we move ahead with his update (might need talk a bit more) it might be worth exploring changing the `<div>` to `<section>` which was omitted to reduce this impact

> * [x]  We know that there are some really useful contextual comments for the design changes [here](https://docs.google.com/document/d/1CG0OUT1tzy7q9juK4O0Iv974854hiBK5GMGF9_0ma08/edit#heading=h.1i6pnax09yoi) - hopefully we could move those into the commit messages to explain the design rationale

This has been done, I've added a lot more detail about the change rationale within the commit from Design(ers) to help outline why the changes were done.

chris-gds commented:
> > * [x]   iOS Voiceover announces the show all button as a main landmark
> 
> I'm 90% certain this due to the testing environment being wrapped in a `main` as this also announced on other elements like breadcrumbs. Will confirm during deeper testing once accordion is placed on a different page.

Testing in situ below, non-issue - no action

https://user-images.githubusercontent.com/71266765/128730749-27674af6-2b7e-460f-8f19-4c75b3d58833.mp4

christopherthomasdesign commented:
Just spotted a couple of small details!

The shape of the circle around the chevron is currently changing between states, so that when the section is open it gets narrower and isn't a perfect circle. You can see in the second image here it looks a bit squished:

<img width="271" alt="Screenshot 2021-08-09 at 16 53 49" src="https://user-images.githubusercontent.com/19834460/128736059-8ffc6f25-9339-4725-b09b-8ce74cdf482d.png">
<img width="269" alt="Screenshot 2021-08-09 at 16 53 54" src="https://user-images.githubusercontent.com/19834460/128736072-55dfb463-d307-4162-834d-91fe0d5d467f.png">


It should stay the same, a perfect circle with only the rotation of the chevron changing.

---

To me, on smaller screens the 'show all sections' looks a little bit far from the sections below. 

<img width="594" alt="Screenshot 2021-08-09 at 16 26 42" src="https://user-images.githubusercontent.com/19834460/128735683-84b1223b-c011-45e8-b8ba-45faa021582c.png">

It currently uses a margin-bottom of 14px. Assuming that's 14 in order to leave 1px for the line, and therefore keep the 5px vertical rhythm. If that's so, I'd probably reduce it to 9px at 'mobile' and back up to 14px for larger screens. Small difference but I think it will tie it together a bit better.

chris-gds commented:
> isn't a perfect circle. You can see in the second image here it looks a bit squished:

Agreed, this new conversion from `pixels` to `em` doesn't quite seem... exact. Will sort.

---

> It currently uses a margin-bottom of 14px. Assuming that's 14 in order to leave 1px for the line, and therefore keep the 5px vertical rhythm.

[taking a look at the Design](https://www.figma.com/file/3yTE9s4EOkvjcqRnqaIzgG/Accordion-Redesign?node-id=343%3A726) it's currently reflective of the 10px across both views however [I've a feeling some adjustments to the Hover state](https://github.com/alphagov/govuk-frontend/pull/2257/commits/b1c57bb81b8121aa90112a8dd6a53b06bffd918c) aren't that obvious.

![Screenshot 2021-08-09 at 17 35 35](https://user-images.githubusercontent.com/71266765/128741615-e65c528a-41ef-4d0b-b6f8-1f87c26fbff8.png)

![Screenshot 2021-08-09 at 17 35 25](https://user-images.githubusercontent.com/71266765/128741617-a5fe1da0-6f2f-47d5-87b3-6cb67619b4ca.png)

>To me, on smaller screens the 'show all sections' looks a little bit far from the sections below.

> If that's so, I'd probably reduce it to 9px at 'mobile' and back up to 14px for larger screens. Small difference but I think it will tie it together a bit better.

In this context I think changing it to a relative value will feel a smidge tighter on mobile and give that 1px 
reduction feel.

chris-gds commented:
As the NVDA fix changes the element to `display:inline` there's less control in styling which results in the summary copy - while focused-  visually displaying a 1px line due the `box-shadow`. A potential fix is to increase the font from `16px` to `17px` however then strays aways from `govuk-body` values.

![Screenshot-2021-08-10-at-18 50 04](https://user-images.githubusercontent.com/71266765/128910085-b7b265ad-35e4-4664-9cd9-70e8494699cd.jpg)

@hannalaakso will confirm with Design
